src/components/Documentation/Markdown/index.tsx

```diff
@@ -50,11 +50,62 @@ const Abbr: React.FC<{ children: [string] }> = ({ children }) => {
   return <Tooltip text={children[0]} />
 }
 
+const Cards: React.FC<{
+  children: Array<object | string>
+}> = ({ children }) => {
+  return (
+    <div className={styles.cards}>
+      {Array.isArray(children)
+        ? children.reduce(
+            (acc: Array<object>, child, i) =>
+              typeof child === 'object'
+                ? [...acc, <div key={i}>{child}</div>]
+                : acc,
+            []
+          )
+        : children}
+    </div>
+  )
+}
+
+const Card: React.FC<{
+  children: Array<object | string> | object | string
+  icon?: string
+  heading?: string
+  headingtag:
+    | string
+    | React.FC<{
+        className: string
+      }>
+}> = ({ children, icon, heading, headingtag: Heading = 'h3' }) => {
+  let iconElement
+  if (Array.isArray(children) && icon) {
+    const firstRealItemIndex = children.findIndex(x => x !== '\n')
+    iconElement = children[firstRealItemIndex]
+    children = children.slice(firstRealItemIndex + 1)
+  }
+  return (
+    <div className={styles.card}>
+      {iconElement && <div className={styles.cardIcon}>{iconElement}</div>}
+      <div className={styles.cardContent}>
+        {heading && <Heading className={styles.cardHeading}>{heading}</Heading>}
+        {children}
+      </div>
+    </div>
+  )
+}
+
 const renderAst = new rehypeReact({
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   createElement: React.createElement as any,
   Fragment: React.Fragment,
-  components: { details: Details, abbr: Abbr, a: Link }
+  components: {
```

Comment:
transformation happens build time, right? I wonder if we can "simplify" the external-links plugin this way?

Reply:
Sort of, but not quite. the HAST is precompiled JSON, and this just map components to json types when rehydrating.
If you're talking about the one that puts a little box next to external links, I do believe we can do that by replacing the `a` here with a `Link` component that has that behavior, and it wouldn't incur any runtime cost beyond normal React.

Reply:
nno, I mean `plugins/external-link` the one that generates embed in blog. Try to grep `external-link` elements in the blog content.

Reply:
okay, I see .. `renderAst` happens client side ... to be honest I don't particularly like this approach, I think `Details` and `Abbr` exist here just because @pavelgrinchenko and @iAdramelk (they can chime in) didn't have enough time to do this properly (e.g. similar to the external-link-plugin) when they were migrating this to Gatsby. Ideally we should have something similar to what we have in the blog where we don't `renderAst` on the client side and just insert html "as-is".

To be honest, I'm not sure how much it would make experience faster and reduce the bundle size - would love to see some experiments. Also, would love to hear from @iAdramelk on this :)

Reply:
My description may have been a bit confusing- the Rehype approach is still turned into HTML by Gatsby like any other page and there isn't a noticeable performance hit. When I say "not build time", I mean these components don't transform or fetch external data like remark/rehype plugins can so it's harder to do things like images (hence the MD image style used for icon boxes)

Compared to `rehype-react`, MDX does move that last little bit of client-side rehydration into the build step, as it completely turns the MD+components into ES5 at the build step.

Reply:
Yeah, you are right that Gatsby renders it statically. And may be I'm wrong, but the question here is on what is happening on navigation/pre-fetch - what is being returned to the client and how is it processed. I remember that in case of blog we have html and use dangerousHtml. In this case does it return JSON (or some "raw" html) that is then being parsed on the client? It means that client needs to pull some libs to do that?

(again, don't have enough time to dig too deep into this - so, it might be I'm completely wrong here)

Reply:
No need to worry, you have a legitimate concern!

Using `rehype-react` at all, which still applies even before the homepage with custom components, does lead to parsing JSON into React for hydration and when running in JS-space. 
However, all the HAST is still transformed, parsed, and ready to be interpreted straight from GraphQL. The `page-data` cache will hold JSON HAST object instead of a raw HTML string.

Even if somehow React doesn't parse the `dangerouslySetInnerHTML` string into components at all and simply inserts it into the DOM, translating fully preprocessed HAST to React is so fast that both methods can be considered equally performant in practical terms.

If I were to make an educated hypothesis, using `html` may be some milliseconds faster to parse and `htmlAst` likely has a bit of a smaller data payload, but the degree is to the point where these are micro-optimizations. I'll try running lighthouse tests on both to see how they actually stack up.

Reply:
Yeah, I'm not worried even about additional cycles we spend on processing, more about JS bundle size. Again, w/o looking into specific data points it's hard to tell exactly, but I remember we decided to keep it all server side for blog (it was built with Gatsby from the very beginning).

At this point, I think it's fine to keep it this way, but eventually I would try to get rid of all libs on the client side (e.g. we have to send Markdown because of tooltips as far as I remember).

Reply:
@shcheklein @rogermparent  Sorry for the long answer. Basically we originally wanted to render all markdown to html on the backend. This would save us some bundle size and calculation time because we don't need to bundle `rehype` and it's plugins to the bundle to process it on the client.

The plan didn't work through. `Abbr` and `Details` components above are not just static html, they require a lot of custom JS logic to work for positioning, interactive stuff, etc. So to make them work we had to use `rehype-react` to replace standard tags with React components. I personally didn't like it very much, but without replacing this components with something a lot simpler, the is no way to make it work on backend.

Card and Cards on the other hand can be implemented using small remark plugin like [external-link-plugin](https://github.com/iterative/dvc.org/blob/master/plugins/external-link-plugin/index.js). It will not reduce bundle size that much, but it will be consistent with the other parts of the app and still a little smaller.

Reply:
@iAdramelk Thanks for the detailed answer! That sheds some light on a few aspects of the project for me that I was unsure of the purpose of.

My personal style and opinion leads me to believe using a Remark plugin for MD custom components when we're already loading and importing `rehype-react` is making a simple feature more complex and fragile for very minimal benefit. I can understand why we do it in the blog's `external-link` since that means the blog pages don't have to bundle `rehype-react`.

I also posted a comment in the related issue- I think the best way forward given the short timeframe until DVC 1.0 drops is to shelve making this page with custom components for a bit and just make an alternate React template for the docs home to present exactly what we want to interested users while requiring the least amount of accommodating changes in our data pipeline.

Once that's handled, we have less of a time crunch to hash out how exactly we want special pages like this to be handled in the future.

Reply:
Resolving this. We have a ticket https://github.com/iterative/dvc.org/issues/1045 for this - Roger please update it to generalize it - that it's not about tooltips only.